src/main/java/org/gridsuite/securityanalysis/server/service/SecurityAnalysisWorkerService.java

```diff
@@ -125,26 +145,75 @@ private static String sanitizeParam(String param) {
-        return f -> f.log(CATEGORY_BROKER_INPUT, Level.FINE)
-                .flatMap(message -> {
-                    SecurityAnalysisResultContext resultContext = SecurityAnalysisResultContext.fromMessage(message, objectMapper);
-
-                    return run(resultContext.getRunContext())
-                            .flatMap(result -> resultRepository.insert(resultContext.getResultUuid(), result)
-                                    .then(resultRepository.insertStatus(resultContext.getResultUuid(), SecurityAnalysisStatus.COMPLETED.name())))
-                            .doOnSuccess(unused -> {
-                                resultPublisherService.publish(resultContext.getResultUuid(), resultContext.getRunContext().getReceiver());
-                                LOGGER.info("Security analysis complete (resultUuid='{}')", resultContext.getResultUuid());
-                            });
-                })
-                .doOnError(throwable -> LOGGER.error(throwable.toString(), throwable))
-                .subscribe();
+    public Consumer<Message<String>> consumeRun() {
+        return message -> {
+
+            SecurityAnalysisResultContext resultContext = SecurityAnalysisResultContext.fromMessage(message, objectMapper);
+            runRequests.add(resultContext.getResultUuid());
+
+            run(resultContext.getRunContext(), resultContext.getResultUuid())
+                    .flatMap(result -> resultRepository.insert(resultContext.getResultUuid(), result)
+                            .then(resultRepository.insertStatus(resultContext.getResultUuid(), SecurityAnalysisStatus.COMPLETED.name()))
+                            .then(Mono.just(result)))
+                    .doOnSuccess(result -> {
+                        if (result != null) {  // result available
+                            resultPublisherService.publish(resultContext.getResultUuid(), resultContext.getRunContext().getReceiver());
+                            LOGGER.info("Security analysis complete (resultUuid='{}')", resultContext.getResultUuid());
+                        } else {  // result not available : stop computation request
+                            if (cancelComputationRequests.get(resultContext.getResultUuid()) != null) {
+                                stoppedPublisherService.publish(resultContext.getResultUuid(), cancelComputationRequests.get(resultContext.getResultUuid()).getReceiver());
+                            }
+                        }
+                    })
+                    .doOnError(throwable -> {
+                        if (!(throwable instanceof CancellationException)) {
+                            LOGGER.error(throwable.toString(), throwable);
+                        }
+                    })
+                    .doFinally(s -> {
+                        futures.remove(resultContext.getResultUuid());
+                        cancelComputationRequests.remove(resultContext.getResultUuid());
```

Comment:
race condition: if the cancel request happens after this but before the runRequests.remove, then the request is not removed from cancelComputationRequests

Reply:
Maybe in another US ?